### PR TITLE
Fix a memory leak in `AutomaticKeepAlive`

### PR DESCRIPTION
**Pull request by GregoryConrad**
Previously, if a KeepAliveHandle outlived an AutomaticKeepAlive (e.g., switching locations in the tree), it'd cause a memory leak by holding onto the AutomaticKeepAlive. This PR fixes that leak by calling `removeListener` once the listener is fired. See more: https://github.com/flutter/flutter/pull/123864#issuecomment-1493365197

This PR was spun off of #123864.

CC @dnfield @Piinks

### CODE REVIEW

**Comment by dnfield** [comment]
Can you please add a test for this to make sure it doesn't get regressed?

**Comment by GregoryConrad** [comment]
@dnfield Should be all set!

**Review by dnfield — APPROVED** [review]
LGTM

**Comment by dnfield** [comment]
Great! This will need a second review from a contributor to land.

**Review by Piinks — APPROVED** [review]
Flutter_LGTM
Thank you! 

I checked the Google testing failure, it is not related to the PR. A rebase should set it straight.

**Comment by goderbauer** [comment]
Did something go wrong with rebasing or updating this branch? Or is it really supposed to touch 158 files?

(removing the auto submit label for now)

**Comment by GregoryConrad** [comment]
@goderbauer Looks like GitHub's online UI messed up the rebase. Same thing happened yesterday. I'll fix it

**Comment by goderbauer** [comment]
Thanks. Re-added the label.

**Comment by GregoryConrad** [comment]
Going to rebase to trigger a re-test. Shouldn't be related to this PR since everything passed before.

**Comment by Piinks** [comment]
Next time, just send a ping. We and re-run most tests individually. The Google testing one is a special one. :)

**Comment by GregoryConrad** [comment]
Ok! Would you remind removing autosubmit for a second? Looks like a rebase somewhere got messed up and changed the engine.version. I'll sort it out

**Comment by GregoryConrad** [comment]
Branch should be fixed & up to date. Sorry about the hassle. Lesson learned: will not be using GitHub's online branch update feature in the future 🙃

**Comment by tirth-multipl** [comment]
> Branch should be fixed & up to date, ready for auto-submit. Sorry about the hassle! Lesson learned: will not be using GitHub's online branch update feature in the future 🙃

you could use this hack to fix 'update rebase (from github) mess' - https://sirolad.medium.com/other-peoples-commit-in-my-pull-request-github-afc1fa7528ea

**Comment by GregoryConrad** [comment]
@Piinks Hi! Is the google testing error still unrelated to this PR? Would a rebase help? Or should it be manually rerun?

Also, this PR should be all set for autosubmit!

**Comment by Piinks** [comment]
Hey @GregoryConrad sorry our CI has been having some troubles the last few days. I've rebased this to see if we can get this landed. There is another change that caused a lot of failures that has been getting caught up in other tests. Thanks for your patience.

**Comment by tirth-multipl** [comment]
> Hey @GregoryConrad sorry our CI has been having some troubles the last few days. I've rebased this to see if we can get this landed. There is another change that caused a lot of failures that has been getting caught up in other tests. Thanks for your patience.

why did number of checks reduced to just 9?!

**Pull request merged**
